### ISSUE TITLE
Ignore Duplicate Scrobbles from Last.fm

### ISSUE THREAD

**Issue by HidYn** [FoxxMD/multi-scrobbler#173]
Firstly, thank you for such a wonderful project. MS has allowed me to focus on ListenBrainz and Maloja instead of Last.fm

I've got what more like a feature request. I use an application called Roon which only supports Last.fm scrobbling and it appears to have a long-standing bug that never gets addressed that involves duplicate scrobbles. https://community.roonlabs.com/t/play-count-not-updating-as-expected/218686/67

![Screenshot_20240723_111517](https://github.com/user-attachments/assets/7067648a-43d3-4290-8d78-776612d496cc)

The track names appear to be slightly different, but the scrobble time is exactly the same down to the second.

I've got Last.fm setup as a source only for Roon and I would love the ability to have those duplicate scrobbles ignored when pulled through to ListenBrainz as at the moment the duplicates get pushed there. It appears to be already ignoring them when pushed through to Maloja to which I have ListenBrainz as the source so no duplicates get posted there.

Keep up the great work.

**Comment by FoxxMD** [comment]
Let me make sure I understand your MS configuration:

```mermaid
flowchart TB

    rn1[Roon] --> rnlfm[Last.fm]

    subgraph multi-scrobbler
    rnlfm[Last.fm Source] --> lzc[Listenbrainz Client]

    lz[Listenbrainz Source] --> ml[Maloja]
    end
```

Is this correct?

___

> I would love the ability to have those duplicate scrobbles ignored 

MS should already be detecting small differences in scrobbles with the same timestamp [as duplicates](https://github.com/FoxxMD/multi-scrobbler/actions/runs/10047314458/job/27768734471?pr=172#step:6:76), I have a whole [test suite](https://github.com/FoxxMD/multi-scrobbler/blob/master/src/backend/tests/scrobbler/scrobblers.test.ts#L92) for it :) Let's try to figure out why this isn't working for your use case.

If you are not already using it please switch to using [File](https://foxxmd.github.io/multi-scrobbler/docs/configuration?configType=file#configuration-types) or [AIO File](https://foxxmd.github.io/multi-scrobbler/docs/configuration?configType=aio#configuration-types) for your [listenbrainz client configuration.](https://foxxmd.github.io/multi-scrobbler/docs/configuration#listenbrainz) Then, add the following `verbose` properties to `options` in your config:

(example of `listenbrainz.json`)
```diff
[
  {
    "name": "brainz",
    "enable": true,
    "configureAs": "client",
    "data": {
      "token": "029b081ba-9156-4pe7-88e5-3be671f5ea2b",
      "username": "MyUSer"
    },
+    "options": {
+     "verbose": {
+        "match": {
+          "onNoMatch": true,
+          "confidenceBreakdown": true
+        }
+      }
+    }
  }
]
```

This will make multi-scrobbler log details to the DEBUG log level when a scrobble client **does not find duplicate scrobbles**, like this:

```
[2024-07-23 09:18:36.986 -0400] DEBUG  : [Test - Test] Source: 9th Wonder / Cordae / Dinner Party / Kamasi Washington / Phoelix / Robert Glasper / Terrace Martin - Freeze Tag @ 2024-07-23T09:17:50-04:00 (S) => Closest Scrobble: 9th Wonder / Cordae / Dinner Party / Kamasi Washington / Phoelix / Robert Glasper / Terrace Martin - Freeze Tag @ 2024-07-23T09:16:49-04:00 (S) => Score 0.70 => No Match
[2024-07-23 09:18:36.986 -0400] DEBUG  : [Test - Test] Breakdown:
Artist: 1.00 * 0.3 = 0.30
Title: 1.00 * 0.4 = 0.40
Time: (No correlation) 0 * 0.5 = 0.00
Time Detail => Existing: 09:16:49-04:00 - Candidate: 09:17:50-04:00 | Temporal Sameness: No correlation | Play Diff: 61s (Needed <60s) | Range Comparison N/A
Score 0.70 => No Match
```

Make sure MS is logging to DEBUG -- if you have not modified env `LOG_LEVEL` or `loglevel` in `config.json` then you are good to go.

Please turn on these settings and restart MS. Then wait for a duplicate scrobble to occur and copy-paste the detailed logs here so we can figure out what is going on.

**Comment by FoxxMD** [comment]
Ignoring visyn configs, it seems you have a circular reference in your configs:


```mermaid
flowchart TB

    subgraph verbose

    lfms[src-hidyn-lastfm] --> lzc[cli-hidyn-listenbrainz]
    lzs[src-hidyn-listenbrainz] --> ml[Maloja]
    lzs --> lfmc[cli-hidyn-lastfm]
%%    lfmc <-.-> lfms
%%    lzs <-.-> lzc
   end

   subgraph simplified
   lfm[hidyn-lastfm] --> lz[hidyn-listenbrainz]
   lz --> mls[Maloja]
   lz --> lfm
   end
```

Is this on purpose? MS isn't really designed/tested for handing this kind of feedback loop. Ideally, all data should flow in one direction.

**Comment by FoxxMD** [comment]
he reason I didn't go this route initially is because Last.fm is awful at handling multi-artist tracks compared to ListenBrainz and Maloja

True, LFM has the least data-rich scrobbling. Why do you need to scrobble back to LFM from LZ? Couldn't it be:

```
Roon -> LFM -> LZ
Pano -> LZ
LZ -> Maloja
```

Or is LFM for recommendations?

P.S. Roon has an API and I've been wanting to [implement it as a source](https://github.com/FoxxMD/multi-scrobbler/issues/83) but I've had a hard time getting the company to communicate. I don't want to pay for it as I won't really be _using_ it, just developing for it.

**Comment by FoxxMD** [comment]
>  You know Roon does a 14 day trial?

Yeah...and its probably enough time to implement a Source but then I need to be able to support users going forward which may require me to reproduce behavior from Roon as well as provide fixes and update the api when they cut new releases. I'd be paying _them_ $15/mo so that I can provide free software to make their product better. It doesn't sit well with me.

> Here's a more expanded log from MS

Ok so here's what I'm parsing from the log (putting here for you but more for future me):

___

LFM source discovers track from history
```
[2024-07-23 15:22:00.216 +0100] INFO   : [App] [Sources] [Lastfm - src-hidyn-lastfm] Discovered => Queens of the Stone Age - Feel Good Hit of the Summer (Album Version) @ 2024-07-23T15:19:06+01:00 (S)
```

LZ client refreshes existing scrobbles (history)
```
[2024-07-23 15:22:00.791 +0100] DEBUG  : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Refreshing recent scrobbles
[2024-07-23 15:22:01.397 +0100] DEBUG  : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Found 40 recent scrobbles
```

LZ client existing scrobbles does not have track, scrobbles
```
[2024-07-23 15:22:01.409 +0100] DEBUG  : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Lastfm: Queens of the Stone Age - Feel Good Hit of the Summer... @ 2024-07-23T15:19:06+01:00 (S) => Closest Scrobble: Queens of the Stone Age - What the Peephole Say @ 2024-07-23T14:00:13+01:00 (S) => Score 0.35 => No Match
[2024-07-23 15:22:01.521 +0100] INFO   : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Scrobbled (Backlog) => (Lastfm) Queens of the Stone Age - Feel Good Hit of the Summer (Album Version) @ 2024-07-23T15:19:06+01:00 (S)
^^^^ Should be storing sent scrobble
```

LZ source discovers track from history
```
[2024-07-23 15:22:35.096 +0100] DEBUG  : [App] [Sources] [Listenbrainz - src-hidyn-listenbrainz] Refreshing recently played
[2024-07-23 15:22:35.753 +0100] INFO   : [App] [Sources] [Listenbrainz - src-hidyn-listenbrainz] Discovered => Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S)
```

Maloja client existing scrobbles does not have track, scrobbles
```
[2024-07-23 15:22:36.495 +0100] INFO   : [App] [Scrobblers] [Maloja - cli-hidyn-maloja] Scrobbled (Backlog)     => (listenbrainz) Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S)
```

LFM client existing scrobbles does not have track, scrobbles _(May be fixed by forced history refresh)_
```
[2024-07-23 15:22:36.779 +0100] INFO   : [App] [Scrobblers] [Lastfm - cli-hidyn-lastfm] Scrobbled (Backlog) => (listenbrainz) Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S)
```

LFM source discovers duplicate track
```
[2024-07-23 15:22:45.975 +0100] DEBUG  : [App] [Sources] [Lastfm - src-hidyn-lastfm] Refreshing recently played
[2024-07-23 15:22:46.293 +0100] INFO   : [App] [Sources] [Lastfm - src-hidyn-lastfm] Discovered => Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S)
```

LZ client does not find duplicate in history _(May be fixed by forced history refresh)_
```
[2024-07-23 15:22:47.540 +0100] DEBUG  : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Lastfm: Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S) => Closest Scrobble: Queens of the Stone Age - What the Peephole Say @ 2024-07-23T14:00:13+01:00 (S) => Score 0.41 => No Match
^^^ Should be checking sent scrobbles, is it not?
```

LZ client scrobbles dup track
```
[2024-07-23 15:22:47.651 +0100] INFO   : [App] [Scrobblers] [Listenbrainz - cli-hidyn-listenbrainz] Scrobbled (Backlog) => (Lastfm) Queens of the Stone Age - Feel Good Hit of the Summer @ 2024-07-23T15:19:06+01:00 (S)
```
___

So there's two things going wrong here, other than the underlying feedback loop:

* scrobble clients should refresh history if track to-be-scrobbled has a timestamp newer than than last history check -- but they are not for some reason
* MS scrobble clients store their own sent scrobbles as well as upstream service scrobbles and (should be) checking both for duplicates -- but this sent scrobbles list isn't triggering dup detection. (This _should_ be compensating for stale upstream scrobble history but isn't)

___

> If you think the feedback loop is the cause I'll ditch it

This would definitely fix the issue but I don't think your scenario is unfixable. The mechanisms listed above should be working to preventing dups even with your set up. I will try to see if I can reproduce in a test suite and determine if these are actually bugs.

In the meantime you can brute-force the refresh history bug by using docker image `foxxmd/multi-scrobbler:pr-174` with the config options [mentioned in this comment](https://github.com/FoxxMD/multi-scrobbler/pull/174#issue-2425421838) to see if this resolving anything for you.

**Comment by FoxxMD** [comment]
Added to clients, same as the verbose logging.

**Comment by FoxxMD** [comment]
Yes that is correct.

P.S. I've made a [post on the roon forums](https://community.roonlabs.com/t/id-like-to-develop-a-scrobbler-application-for-roon/278031) as a final effort to try to get an account for dev.

**Comment by FoxxMD** [comment]
Great! So forcing a refresh of the client's upstream scrobbles is preventing dup scrobbles when the title is very similar.

We can see the reason the LFM duplicate happened is because MS didn't think the track title's were similar enough:

```
[2024-07-23 19:44:28.972 +0100] DEBUG  : [App] [Scrobblers] [Lastfm - cli-hidyn-lastfm] Listenbrainz: The Starting Line - The B-List @ 2024-07-23T19:38:54+01:00 (S) => Closest Scrobble: The Starting Line - The B-List (Album Version) @ 2024-07-23T19:38:54+01:00 (S) => Score 0.91 => No Match
[2024-07-23 19:44:28.973 +0100] DEBUG  : [App] [Scrobblers] [Lastfm - cli-hidyn-lastfm] Breakdown:
Artist: 1.00 * 0.3 = 0.30
Title: 0.27 * 0.4 = 0.11
Time: (Exact) 1 * 0.5 = 0.50
Time Detail => Existing: 19:38:54+01:00 - Candidate: 19:38:54+01:00 | Temporal Sameness: Exact | Play Diff: 0s (Needed <10s) | Range Comparison N/A
Score 0.91 => No Match
```

```
Candidate           => The Starting Line - The B-List                 @ 2024-07-23T19:38:54+01:00
Closest existing    => The Starting Line - The B-List (Album Version) @ 2024-07-23T19:38:54+01:00
```

* Artist similarity 1.00 = 100%
* Time similarity 1 = 100%
* Title similarity 0.27 = 27%

With a weighted score of 0.91 it _almost_ considered it a dupe (needs to be `>= 1`) but the general heuristic for finding dupes isn't wrong here. The title is definitely dissimilar enough that, generally, it shouldn't count this as dups (even though we we know it is).
___

Ironically, LFM actually also provides metabrainz IDs for historic scrobbles but only for artist and album -- no release id. So it's not possible to programmatically determine it's the same track without using the track title at face value.

The score weights are pretty well balanced from my testing so I'm reluctant to change those. I also don't want to hardcode an exception for `(Album Version)` as this may very well be a legitimately different track in another scenario.

My first impression for fixing this, then, would be to allow user-configrable terms to be removed from a play so that edge cases like this can be taken care of by the user in `listenbrainz.json` (or maybe the source config?) I'll need to think on how to best go about this that doesn't introduce too much complexity.

**Comment by FoxxMD** [comment]
@HidYn please update to the latest `foxxmd/multi-scrobbler:pr-174` image when you can. The `refreshForce` option is gone but some new code I added may fix the stale refresh data without it.

If you find it is still doing duplicates that **aren't** related to the track title being different you can use the new option `refreshStaleAfter` instead of `refreshForce` like this:

```
    "options": {
     "refreshStaleAfter": 3,
     "verbose": {
        "match": {
          "onNoMatch": true,
          "confidenceBreakdown": true
        }
      }
    }
```

`refreshStaleAfter` forces the client to refresh scrobbled plays from upstream service if last refresh was at least X seconds ago. Setting it to `0` would effectively be `refreshForce: true`

**Comment by FoxxMD** [comment]
Nah it's all good I've got a test written for it already anyways. Thanks again for the detailed writeup, logging, and testing.

**Comment by FoxxMD** [comment]
@HidYn if you'd like to try this out and go back to your original setup I have a way to modify scrobble data using configs. Switch to `foxxmd/multi-scrobbler:develop` image.

The docs are [here](https://github.com/FoxxMD/multi-scrobbler/blob/develop/docsite/docs/configuration/transforms.mdx) until it's cut for a release and for your [particular scenario](https://github.com/FoxxMD/multi-scrobbler/issues/173#issuecomment-2246133922) the config would look like this for `lastfm.json`:

```json5
[
  {
    "name": "myLastFm",
    "configureAs": "client"
    // ...
    "options": {
      "refreshStaleAfter": 3,
      // ...
      "playTransform": {
        "compare": {
          "existing": {
            "title": [
              "(Album Version)" // ignore '(Album Version)' in existing scrobble titles during duplicate comparison
            ]
          },
        },
      }
    }
  }
]
```

**Comment by FoxxMD** [comment]
Scrobble modification and refresh forcing configuration (`refreshStaleAfter`) have been [released in `v0.8.2`.](https://github.com/FoxxMD/multi-scrobbler/releases/tag/0.8.2)
